services/repository-managers/src/main/java/org/sagebionetworks/repo/manager/schema/EntitySchemaValidator.java

```diff
@@ -24,45 +29,65 @@ public class EntitySchemaValidator implements ObjectSchemaValidator {
 	private final JsonSchemaValidationManager jsonSchemaValidationManager;
 	private final SchemaValidationResultDao schemaValidationResultDao;
 	private final DerivedAnnotationDao derivedAnnotationDao;
-
+	private final TransactionalMessenger messenger;
+	
 	@Autowired
 	public EntitySchemaValidator(EntityManager entityManger, JsonSchemaManager jsonSchemaManager,
 			JsonSchemaValidationManager jsonSchemaValidationManager,
-			SchemaValidationResultDao schemaValidationResultDao, DerivedAnnotationDao derivedAnnotationDao) {
+			SchemaValidationResultDao schemaValidationResultDao, DerivedAnnotationDao derivedAnnotationDao,
+			TransactionalMessenger messenger) {
 		super();
 		this.entityManger = entityManger;
 		this.jsonSchemaManager = jsonSchemaManager;
 		this.jsonSchemaValidationManager = jsonSchemaValidationManager;
 		this.schemaValidationResultDao = schemaValidationResultDao;
 		this.derivedAnnotationDao = derivedAnnotationDao;
+		this.messenger = messenger;
 	}
 
 	@WriteTransaction
 	@Override
 	public void validateObject(String entityId) {
 		ValidateArgument.required(entityId, "entityId");
+		
+		boolean sendEntityUpdate = false;
+		
 		try {
 			JsonSchemaObjectBinding binding = entityManger.getBoundSchema(entityId);
 			JsonSubject entitySubject = entityManger.getEntityJsonSubject(entityId, false);
 			JsonSchema validationSchema = jsonSchemaManager
 					.getValidationSchema(binding.getJsonSchemaVersionInfo().get$id());
 			ValidationResults results = jsonSchemaValidationManager.validate(validationSchema, entitySubject);
 			schemaValidationResultDao.createOrUpdateResults(results);
-			
-			if(binding.getEnableDerivedAnnotations()) {
+		
+			if (binding.getEnableDerivedAnnotations()) {
 				Optional<Annotations> annoOption = jsonSchemaValidationManager.calculateDerivedAnnotations(validationSchema, entitySubject.toJson());
 				if(annoOption.isPresent()) {
 					derivedAnnotationDao.saveDerivedAnnotations(entityId, annoOption.get());
-				}else {
-					derivedAnnotationDao.clearDerivedAnnotations(entityId);
+					sendEntityUpdate = true;
+				} else {
+					sendEntityUpdate = derivedAnnotationDao.clearDerivedAnnotations(entityId);
 				}
-			}else {
-				derivedAnnotationDao.clearDerivedAnnotations(entityId);
+			} else {
+				sendEntityUpdate = derivedAnnotationDao.clearDerivedAnnotations(entityId);
 			}
 
 		} catch (NotFoundException e) {
 			schemaValidationResultDao.clearResults(entityId, ObjectType.entity);
-			derivedAnnotationDao.clearDerivedAnnotations(entityId);
+			sendEntityUpdate = derivedAnnotationDao.clearDerivedAnnotations(entityId);
+		}
+		
+		if (sendEntityUpdate) {
+			// When the derived annotations are computed we want to trigger the replication so that the index is updated, since this is not an actual update of the entity
+			// we simply publish a non-migratable change that is processed as a normal change message
+			messenger.publishMessageAfterCommit(
```

Comment:
publishMesageAferCommit should ensure both the user and the change number are set, so callers are not required to pass that info.